### ISSUE TITLE
appsodyapplication question re:  dev.local and docker images

### ISSUE THREAD

**Issue by kewegner** [appsody/appsody-operator#87]
@chilanti @neeraj-laad

Environment:
Appsody cli .0.4.1
Docker Desktop
macOs 10.14.6

In the Appsody CLI if we do 'appsody deploy' from our appsody project directory 'xyz'.
The code creates a docker image called 'xyz' which is available only in the local docker desktop, it has not been pushed to a remote repository.

The operator is able to create a working application even though the image is not tagged as 
'dev.local/xyz'

We cleaned up our k8s from our initial testing:

We then tried to add: 

createKnativeService: true 

to the app-deploy.yaml for 'xyz' project.

We found that the operator created a KnativeService, but that it fails to come up:

- lastTransitionTime: "2019-09-04T19:14:16Z"
   message: 'Revision "xyz-v78rb" failed with message: Unable to fetch image "xyz":
     UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:library/xyz
     Type:repository]].'
   reason: RevisionFailed

We then did the following:

appsody deploy -t dev.local/xyz

--where -t tags the image in docker as specified.

And our KnativeService deployed successfully.

The difference is that the docker image has the 'dev.local/' in the tag which appears in the KnativeService definition when running a kubectl query.  If we don't specify the -t option, then the image name in both the ksvc definition (and the docker image name) is 'xyz' producing the error above.

Our question is why did the non Knative path for the operator/appsodyapplication not require the 'dev.local/' prefix for the docker image?

**Comment by arturdzm** [comment]
Regular deployments will use locally stored image, although if it is not found the default usually is to try to pull it from docker hub. Knative has a feature built in where I think for purpose of differentiating images with the same tag it tries to request the image digest (sha) from image registry before creating a deployment/pods, since there is no registry specified it might be assuming docker hub or whichever registry is a default

**Comment by chilanti** [comment]
Arthur is correct. 
TL;DR
The reason it works with "straight" operator deployment (non-Knative) on a local Docker-for-desktop cluster is that the cluster just pulls the image from the local Docker cache (or - if that wasn't found there, would attempt to pull it from Dockerhub). Knative tries to retrieve a digest - and in order to do that it has to authenticate to the docker registry where it thinks it can find the image. This is why there's a configuration map where one can specify to skip the retrieval of the digest through an appropriate parm (`registriesSkippingTagResolving` as described [here](https://knative.dev/docs/serving/tag-resolution/).)
In the original Knative deployment path of `appsody deploy`, the CLI names the image explicitly with a prefix `dev.local` - under the assumption that the developer has set `registriesSkippingTagResolving: "dev.local" ` in the config map, as suggested in our docs. 
But the new path to Knative deployment doesn't do that. 

**Bottom line**
I think we need to add the logic that names the image appropriately in the CLI in the case of Knative deployment - regardless of whether the deployment occurs via the operator or not.

**Comment by chilanti** [comment]
I would propose closing this and opening a bug on appsody/appsody.

**Comment by kewegner** [comment]
Closed in reference to:
Appsody deploy with Knative option fails to prefix the image name #281
